### PR TITLE
Add extra dictionary support

### PR DESCRIPTION
**Pull request by cdchapman**
**_Requires at least hunspell version 1.3.4_** see [the commit](https://github.com/hunspell/hunspell/commit/ee3bece2e6f92a03ceadd157fb51056a3bb9bf2d) when this interface was introduced. I don't know whether there is a way to attach the function only for specific versions of hunspell.

### CODE REVIEW

**Comment by cdchapman** [comment]
A [similar situation](http://ffi.github.io/ruby-ffi-archive/messages/20110904-%5Bruby-ffi%5D%20attach_function%20a%20bit%20draconian-244.html) and [a workaround](http://ffi.github.io/ruby-ffi-archive/messages/20110904-%5Bruby-ffi%5D%20attach_function%20a%20bit%20draconian-244.html) came up in an early `ruby-ffi` discussion. I'll add another commit to use this approach.

**Comment by postmodern** [comment]
When FFI attempts to bind a function that doesn't exist, it raises a `FFI::NotFoundError` exception. You could catch that when attaching the additional functions, and fail silently?

**Comment by Envek** [comment]
@cdchapman, thank you, this is exactly functionality that I need!
@postmodern, what should be done yet in this PR to get it merged?

One more idea for ease of use is to make `add_dic` to search for dictionaries in `Hunspell.directories` as `FFI::Hunspell::Dict.open` method does to ease of adding other languages dictionaries.

My use case: I want to check words with russian, english and my custom dictionary (with my domain words).

For now with the code from this pull request I can do it as:

```rb
require 'ffi/hunspell'
dict = FFI::Hunspell.dict('ru_RU')
dict.add_dic('/usr/share/hunspell/en_US.dic') # ← I want it to be just en_US, I don't want to stick to any one path (which is differs between distros, etc)
dict.add_dic('/home/envek/evl.ms/ebay.ru/misc/hunspell.dic') # dxg/MS is here

# Usage:
dict.check?('собака'.encode(dict.encoding)) # => true
dict.check?('dog'.encode(dict.encoding)) # => true
dict.check?('dxg'.encode(dict.encoding)) # => true
```

I can continue work on this PR with your permission. WDYT?

File: lib/ffi/hunspell/hunspell.rb [hunk 1]
```diff
@@ -28,7 +28,14 @@ module Hunspell
     begin
       attach_function :Hunspell_add_dic, [:pointer, :string], :int
     rescue FFI::NotFoundError
-      warn("Hunspell_add_dic was not found in [#{ffi_libraries.map(&:name).join(", ")}]. You must to install Hunspell 1.3.4 or later if you need this functionality.")
+      define_singleton_method :method_missing do |symbol, *arguments, &block|
```
Comment (postmodern) [inline lib/ffi/hunspell/hunspell.rb]: I would just define `Hunspell_add_dic` and have it raise `NotImplementedError`.

**Review by postmodern — CHANGES_REQUESTED** [review]
Just define a place holder method and avoid any `method_missing` magic.

**Comment by postmodern** [comment]
Sorry for the delay. Only one minor issue, but this looks ready to be merged.

**Comment by cdchapman** [comment]
Hi @Envek. Sorry for the delay. It would not make sense to use the russian affix file, for example, with the `en_US` dictionary because the affixes for each language mean different things. Extra dictionaries reuse the affix file from the main dictionary. See hunspell/hunspell#348.

A better approach would be the following:

```ruby
require 'ffi/hunspell'
dict = FFI::Hunspell.dict('ru_RU')
dict = dict.add_dic('/full/path/to/russian_medical_words.dic') #

english_dict = FFI::Hunspell.dict('en_US')
english_dict.add_dic('/full/path/to/custom.dic') # dxg/MS is here

# Usage:
dict.check?('собака'.encode(dict.encoding)) # => true
dict.check?('полиглактином'.encode(dict.encoding)) # => true
english_dict.check?('dog'.encode(dict.encoding)) # => true
english_dict.check?('dxg'.encode(dict.encoding)) # => true
```
It is useful anyhow to know what language is being checked. An application could guess using some sort of heuristics, but it is good to be explicit about the guessing because it may choose the wrong dictionary.

**Comment by Envek** [comment]
Yes, I figured that and finally I did exactly this: have two main dictionaries for every language and two additional dictionaries (one for each language).

But anyway this PR is still needed.

**Comment by postmodern** [comment]
Been busy at work. Finally got around to working on ffi-hunspell. Merged and will be in 0.6.0.

**Comment by wynksaiddestroy** [comment]
Is it possible by chance to release a new version to offer this new feature?

**Pull request merged**